enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,350 @@
+
+**End User** is a human user operating an Edge Device.
+
+1. Device Administrator releases an updated system image over the network to
+upgrade devices in the field
+2. Edge Devices monitor system upgrades and download them in the background
+3. Edge Device upgrade staged on `rpm-ostree` does not take place until the next
+system reboot
+
+> In step 3, the system reboot may be scheduled by a Device Administrator, or
+> spontaneous due to a End User action, power failure, etc.
+
+1. End Users continue using the current software version until their devices are
+rebooted
+2. Upon a device reboot, End Users get a new software version only if it has been
+downloaded and staged
+3. End Users wait until the new software verification procedure completes
+4. Depending on the verification results, End Users get the updated software,
+or the upgrade is rolled back
+
+> In step 3, the verification procedure may reboot a device multiple times until
+> the upgrade is declared successful or failed.
+
+Once the system is rebooted implicitly (i.e. `cron` schedule) or explicitly
+by a Device Administrator, End User, etc.,
+the [greenboot-healthcheck.service](https://github.com/fedora-iot/greenboot#how-does-it-work)
+runs to verify the system validity.
+
+If the system validity procedures are successful, the version of the software
+is declared valid for end users. Otherwise, `greenboot` enters a system
+verification loop by repeating reboots up-to `GREENBOOT_MAX_BOOT_ATTEMPTS` times
+before attempting to rollback to the previous version of the software.
+
+```mermaid
+graph TD
+    A[Software vN] --> |Software Upgrade| B[Reboot Pending]
+    B --> |System Reboot| D{Validate Software vN+1}
+    D --> |Valid| E[Software vN+1]
+    D --> |Invalid| F[Verification Loop]
+    F --> |Up to Max Reboots|D
+    F --> |All Failed Attempts| G[Rollback]
+    G --> A
+```
+
+### API Extensions
+
+None
+
+### Implementation Details
+
+The MicroShift core RPM package will include the `microshift-greenboot-required.sh`
+health check script to validate that all the required MicroShift services are
+up and running. There will be no explicit dependency on the `greenboot` RPM.
+The script will be installed into the `/etc/greenboot/check/required.d` directory
+and it will not be executed during the system boot in case the `greenboot` package
+is not present.
+
+> If other scripts are present in the `/etc/greenboot/check/required.d` directory,
+> they will be executed by `greenboot` in the alphanumeric order.
+
+The script performs the following sanity checks:
+* Exit if the current user is not `root`
+* Exit if MicroShift service is not enabled
+* Wait for the MicroShift service to become active
+
+If the `greenboot-default-health-check` RPM subpackage is installed, it already
+includes [health check scripts](https://github.com/fedora-iot/greenboot#health-checks-included-with-subpackage-greenboot-default-health-checks)
+verifying that DNS and `ostree` services can be accessed.
```

Comment:
Note this package is also optional, because - depending on scenario - there may not be network (at update time or ever), so testing for DNS or `ostree` server would produce a false positive.

Reply:
Yes, the section starts with "if the package is installed". Should I make is more explcit somehow?

Reply:
No action needed here, no. I just wanted to point this out to say that MicroShift's health-checks need to be in a separate RPM, too, otherwise we cannot make them optional.